### PR TITLE
#13332: add ttnn implementation for Bert-Tiny model

### PR DESCRIPTION
**Pull request by vigneshkeerthivasanx**
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/13332)

### Problem description
Add ttnn implementation for BERT-Tiny model.

### What's changed
ttnn support is enabled for BERT-Tiny model along with demo, e2e device and test perf.

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable)
- [x] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by ayerofieiev-tt — COMMENTED** [review]
Ignore the below, found the script
-It does not look like we added this model to CI-

**Review by cfjchu — APPROVED** [review]

**Comment by uaydonat** [comment]
We need to see the Actions that contain the newly added CI tests to be launched and link posted here. The tests have to pass before the PR is approved.

See an example here: https://github.com/tenstorrent/tt-metal/pull/15005#issue-2655932445

**Comment by tt-rkim** [comment]
Please post your passing links.

I believe your integration tests is causing issues with distilbert in nightly FD.

**Comment by vigneshkeerthivasanx** [comment]
> Please post your passing links.
> 
> I believe your integration tests is causing issues with distilbert in nightly FD.

Will trigger CIs again and post passing links

File: models/demos/bert_tiny/tests/test_performance.py
Comment (uaydonat) [inline models/demos/bert_tiny/tests/test_performance.py:102]: tests should fail if performance is lower than expected

File: models/demos/bert_tiny/demo/demo.py
Comment (uaydonat) [inline models/demos/bert_tiny/demo/demo.py:143]: test should fail if answers are wrong or perf slower than expected

File: models/demos/bert_tiny/tests/test_performance.py
Comment (uaydonat) [inline models/demos/bert_tiny/tests/test_performance.py:129]: test should fail if perf is slower than expected

File: tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py
Comment (uaydonat) [inline tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py:109]: does this test run anywhere?

File: models/demos/bert_tiny/tests/test_performance.py
Comment (vigneshkeerthivasanx) [reply models/demos/bert_tiny/tests/test_performance.py:102]: Updated the file with assert statements.

File: models/demos/bert_tiny/demo/demo.py
Comment (vigneshkeerthivasanx) [reply models/demos/bert_tiny/demo/demo.py:143]: Bert-Tiny model results are not promising but got approval from @mbahnasTT for the demo results.

File: models/demos/bert_tiny/tests/test_performance.py
Comment (vigneshkeerthivasanx) [reply models/demos/bert_tiny/tests/test_performance.py:133]: Device perf will fail by default if perf is slower than expected.

File: tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py
Comment (vigneshkeerthivasanx) [reply tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py:109]: This is a sub-module test for Bert-Tiny model and this is tested in CIs.

**Comment by tt-rkim** [comment]
There have been some instability issues with GS device perf. I have rebooted the machine

**Comment by vigneshkeerthivasanx** [comment]
All post-commit tests : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/11948527871) (All Passed)
Nightly fast dispatch tests : [n150](https://github.com/tenstorrent/tt-metal/actions/runs/11948529434/job/33316480128#step:8:3975) , [n300](https://github.com/tenstorrent/tt-metal/actions/runs/11948529434/job/33316481029#step:8:3994),  [e150](https://github.com/tenstorrent/tt-metal/actions/runs/11948529434/job/33316479151) (Bert-Tiny is passing)
(Single-card) Demo tests :  [n150](https://github.com/tenstorrent/tt-metal/actions/runs/11948530844/job/33316584082#step:9:1076) (Bert-Tiny is passing), n300  (Bert-Tiny is passing)
(Single-card) Device perf regressions : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/11948536931) (GS failing with test hang)
(Single-card) Model perf tests : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/11948539715) (All passed)

File: models/demos/bert_tiny/tests/test_performance.py
Comment (uaydonat) [reply models/demos/bert_tiny/tests/test_performance.py:133]: No, check_device_perf has assert_on_fail=False by default, you need to set it to True.

File: tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py
Comment (uaydonat) [reply tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py:109]: I see test_performance is calling test_bert_tiny.py::test_bert_for_question_answering but I cannot find any where this is called. Maybe I am missing it? Where is call for this test?

File: models/demos/bert_tiny/tests/test_performance.py
Comment (vigneshkeerthivasanx) [reply models/demos/bert_tiny/tests/test_performance.py]: Updated the code with assert_on_fail=True

File: models/demos/bert_tiny/demo/demo.py
Comment (vigneshkeerthivasanx) [reply models/demos/bert_tiny/demo/demo.py:143]: Pytorch results does not align with the dataset for most of the cases.
However, Pytorch and TTNN results match in most cases, and the results can be found here: [Link](https://docs.google.com/spreadsheets/d/1dBKFjR5tvsdn3oCvyMyOX1D_z_LwMWEI-q7MSQjvo_o/edit?gid=1101027898#gid=1101027898)

Accuracy between TTNN model and SQUADv2 reference:
```
Eval: exact: 14.423076923076923 
F1: 22.086671341479036
```

Accuracy between PyTorch model and SQUADv2 reference:
```
exact: 19.23076923076923 
F1: 24.28325947556717
```

Accuracy between TTNN and PyTorch model:
```
exact: 58.65384615384615
F1: 0.5875457875457875
```

File: tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py
Comment (vigneshkeerthivasanx) [reply tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py:109]: test_bert_tiny.py::test_bert_for_question_answering is present in the below mentioned path [tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py](https://github.com/tenstorrent/tt-metal/compare/main...vignesh/ttnn_bert_tiny#diff-0a8c71d42b9f83650b262e8804fa85466fda47861d90a4e841eaaa7ab04f7bfbR192)

**Review by tt-rkim — APPROVED** [review]

File: tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py
Comment (uaydonat) [reply tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py:109]: I am asking which function is calling `test_bert_output_inference`. Does this test run anywhere?

**Review by uaydonat — APPROVED** [review]

File: tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py
Comment (vigneshkeerthivasanx) [reply tests/ttnn/integration_tests/bert_tiny/test_bert_tiny.py:109]: No, this isn't running anywhere; it's simply a test to test that specific sub-module.

**Comment by vigneshkeerthivasanx** [comment]
CI links
All post-commit tests: [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12004716771) (PASSED)
Nightly fast dispatch tests : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12004718436) (Bert_tiny model tests passed)
(Single-card) Demo tests : [n150](https://github.com/tenstorrent/tt-metal/actions/runs/12004719754/job/33472194514) (PASSED), [n300](https://github.com/tenstorrent/tt-metal/actions/runs/12004719754/job/33472195132) (PASSED),
(Single-card) Device perf regressions : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12004721520) (PASSED)
(Single-card) Model perf tests : [Link](https://github.com/tenstorrent/tt-metal/actions/runs/12004743276) (Bert_tiny model tests passed)

Can this be merged into the main?

**Comment by tt-rkim** [comment]
Yes , you may squash and merge

**Pull request merged**
